### PR TITLE
asking as an independent command

### PR DESCRIPTION
**Pull request by sjpotter**
Based on conversation the understanding of asking was incorrect, its an independent command that sets a flag on the client, this is to make it the same.

- saves client state when go into asking mode
- set RedisRaftCommandArray as asking if client put itself into asking state
- no longer have to parse multi specially
- serialize/deserialize the asking bool the log entry for RaftRedisCommandArray
- no longer parse asking in the apply path (as not appended to log as a command)

### CODE REVIEW

File: src/raft.c [hunk 1]
```diff
@@ -1138,8 +1136,9 @@ RRStatus RedisRaftInit(RedisModuleCtx *ctx, RedisRaftCtx *rr, RedisRaftConfig *c
         rr->resp_call_fmt = "v";
     }
 
-    /* Client state for MULTI support */
+    /* Client state for MULTI/ASKING support */
     rr->multi_client_state = RedisModule_CreateDict(ctx);
```
Comment (sjpotter) [inline src/raft.c:1140]: combine into a single client_state dict that is a struct that is expandable.

**Comment by sjpotter** [comment]
Q. re should asking be serialized as a bit/flag or should it be appended to the array and then parsed out on apply time.

File: src/redisraft.h [hunk 2]
```diff
@@ -700,6 +701,17 @@ typedef struct JoinLinkState {
...
+typedef struct MultiState {
+    RaftRedisCommandArray cmds;
+    bool active;
+    bool error;
+} MultiState;
+
+typedef struct ClientState {
+    MultiState multiState;
+    bool asking;
+} ClientState;
```
Comment (tezc) [inline src/redisraft.h:713]: I feel like we'll use client struct a lot in future, I'd use a shorter name, just "struct Client" probably.

File: src/redisraft.h [hunk 3]
```diff
@@ -700,6 +701,17 @@ typedef struct JoinLinkState {
...
 } JoinLinkState;
 
+typedef struct MultiState {
+    RaftRedisCommandArray cmds;
+    bool active;
+    bool error;
+} MultiState;
+
+typedef struct ClientState {
+    MultiState multiState;
```
Comment (tezc) [inline src/redisraft.h:711]: Just for style, can you use `multi_state` as variable name.

File: src/redisraft.h [hunk 4]
```diff
@@ -778,6 +790,10 @@ void FreeImportKeys(ImportKeys *target);
 unsigned int keyHashSlot(const char *key, size_t keylen);
 unsigned int keyHashSlotRedisString(RedisModuleString *s);
 RRStatus parseHashSlots(char * slots, char * string);
+ClientState * GetClientState(RedisRaftCtx *rr, RedisModuleCtx * ctx);
```
Comment (tezc) [inline src/redisraft.h:793]: Regarding naming functions, I did many mistakes and merged bunch of things previously but trying to follow a style recently. So, for exported functions, function names start with uppercase letter and with a common prefix. So, maybe these functions should start with `Client..()`

File: src/util.c [hunk 5]
```diff
@@ -500,4 +500,34 @@ RRStatus parseHashSlots(char * slots, char * string)
...
+{
+    ClientState *clientState = RedisModule_Calloc(sizeof(ClientState), 1);
+    RedisModule_DictSetC(rr->client_state, &client_id, sizeof(client_id), clientState);
+}
+
+void FreeClientState(RedisRaftCtx *rr, unsigned long long client_id)
+{
+    ClientState *state = NULL;
+
+    if (RedisModule_DictDelC(rr->client_state, &client_id,
```
Comment (tezc) [inline src/util.c:526]: I think if we don't have a client struct here, we are in trouble (it shouldn't be possible?). Maybe, we should just assert here. We can also use assert when we add a client to dict, just to verify there is no existing client.

File: src/redisraft.c [hunk 6]
```diff
@@ -1716,7 +1774,7 @@ static void handleInfo(RedisModuleInfoCtx *ctx, int for_crash_report)
     RedisModule_InfoAddFieldULongLong(ctx, "snapshots_created", rr->snapshots_created);
 
     RedisModule_InfoAddSection(ctx, "clients");
-    RedisModule_InfoAddFieldULongLong(ctx, "clients_in_multi_state", MultiClientStateCount(rr));
+    RedisModule_InfoAddFieldULongLong(ctx, "clients_state_tracking", ClientStateCount(rr));
```
Comment (tezc) [inline src/redisraft.c:1777]: Right now, this field becomes "connected client count". So, if I'm not mistaken, this will be same with `connected_clients` field in INFO. 

Previously, it was "clients in multi state" and a test uses it. Maybe, we loop over clients and count clients in multi state ?

If we can run/verify test in some other way, it might also mean we don't need this field. @yossigo Do you think this is useful? Redis doesn't report clients in multi state if I'm not mistaken.

File: src/multi.c [hunk 7]
```diff
@@ -35,39 +35,17 @@
...
-    bool error;
-} MultiState;
-
-void MultiInitClientState(RedisRaftCtx *rr)
-{
-    rr->multi_client_state = RedisModule_CreateDict(NULL);
-}
-
-uint64_t MultiClientStateCount(RedisRaftCtx *rr)
+void ResetMultiClientState(ClientState * clientState)
```
Comment (tezc) [inline src/multi.c:38]: Looks like multi.c is not necessary anymore (it's getting shorter). If you want, we can delete the file and move the function here back into the redisraft.c.

File: src/redisraft.c [hunk 8]
```diff
@@ -549,16 +549,65 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
     return false;
 }
 
+static bool AskingHandleCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, RaftRedisCommandArray *cmds)
```
Comment (tezc) [inline src/redisraft.c:552]: I added a comment about style and I know it's not consistent in the codebase (I did bunch of mistakes) but let's try to follow the common style:
- Exported functions start with uppercase letter and a common prefix: 
   - RaftLogCreate(), RaftLogOpen() etc.
   - NodeAddrParse(), NodeAddrEqual() etc.
- Static functions start with lowercase letter. 
- Exception to this: common.c functions start with lowercase. (That's just the reality). So, these are just short helper functions . Probably, it's okay we keep it that way, start with lowercase letter without a common prefix as they are unrelated with each other anyway.

File: src/redisraft.c [hunk 9]
```diff
@@ -604,6 +653,8 @@ static void handleRedisCommand(RedisRaftCtx *rr,
         if (rr->config->quorum_reads) {
             RaftReq *req = RaftReqInit(ctx, RR_REDISCOMMAND);
             RaftRedisCommandArrayMove(&req->r.redis.cmds, cmds);
+            /* need to save asking flag when we move the cmds */
+            req->r.redis.cmds.asking = cmds->asking;
```
Comment (tezc) [inline src/redisraft.c:657]: Maybe this should be part of RaftRedisCommandArrayMove(). Is there an exception to this? Whenever we move a command into the array, we also adopt asking state?

File: src/redisraft.h (hunk 3 above)
Comment (sjpotter) [reply src/redisraft.h:711]: ok

File: src/redisraft.h (hunk 2 above)
Comment (sjpotter) [reply src/redisraft.h:713]: I personally like ClientState, I think the little bit more descriptive is fine.

File: src/redisraft.h (hunk 4 above)
Comment (sjpotter) [reply src/redisraft.h:793]: hmm.  So we want SubjectVerb not VerbSubject?

File: src/redisraft.h (hunk 2 above)
Comment (tezc) [reply src/redisraft.h:713]: ok

File: src/util.c (hunk 5 above)
Comment (sjpotter) [reply src/util.c:526]: sure, its validating redis, which is fine.

File: src/redisraft.h (hunk 4 above)
Comment (tezc) [reply src/redisraft.h]: Normally, in C world, people use a common prefix in exported function names (C version of namespaces). So, I guess it naturally becomes SubjectVerb because of the prefix. Good thing is that when you see that function name, you know where to find it. If a function starts with Shard... I know it's in cluster.c or with Conn.. it's in connection.c etc.

File: src/multi.c (hunk 7 above)
Comment (sjpotter) [reply src/multi.c:37]: So I was thinking sort of the opposite, it be nicer if different commands were handled in different files.

File: src/redisraft.c (hunk 8 above)
Comment (sjpotter) [reply src/redisraft.c:552]: in this case, isn't your comment consistent with this?

File: src/redisraft.c (hunk 9 above)
Comment (sjpotter) [reply src/redisraft.c:657]: possibly.  we dont want to overwrite state each time from true to false (I think) thought that could be solved via a cmds.asking |= cmds->asking within it.

File: src/redisraft.c (hunk 8 above)
Comment (tezc) [reply src/redisraft.c:552]: Sorry I wasn't clear.  Yeah both are same.  The other comment(https://github.com/RedisLabs/redisraft/pull/347#discussion_r889877447) is consistent with this.

File: src/redisraft.c (hunk 9 above)
Comment (tezc) [reply src/redisraft.c:657]: > we dont want to overwrite state each time from true to false

Not sure I understand. What is the problem here? Also, we are going to set asking true once because if client is in asking mode, it will stay in asking mode forever, right?  So, once we are in asking mode, we'll never get "asking=false", no? (We may need to validate this somewhere though).

File: src/multi.c (hunk 7 above)
Comment (tezc) [reply src/multi.c:37]: In general, I find harder to read/navigate multiple files, especially on github.  (Also, it reminds me my Java days that you have billion of files even in the simplest project, [link](https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition)). 

If you have bunch of commands that you handle in 10 lines of code, creating a file for each seems overkill to me but anyway, I don't have a strong opinion.

File: src/redisraft.c [hunk 10]
```diff
@@ -549,11 +549,60 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
+    ClientState *clientState = ClientStateGet(rr, ctx);
+    return clientState->asking;
+}
+
 static void handleRedisCommand(RedisRaftCtx *rr,
                                RedisModuleCtx *ctx, RaftRedisCommandArray *cmds)
 {
     Node *leader_proxy = NULL;
 
+    if (AskingHandleCommand(rr, ctx, cmds)) {
```
Comment (tezc) [inline src/redisraft.c:598]:
```suggestion
    if (handleAsking(rr, ctx, cmds)) {
```

Style. Local functions start with lowercase letter.

File: src/redisraft.c [hunk 11]
```diff
@@ -549,11 +549,60 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
+        return;
+    }
+
+    if (GetAskingState(rr, ctx)) {
```
Comment (tezc) [inline src/redisraft.c:602]: Probably no need this function anymore and it can be part of handleAsking() above.

File: src/redisraft.c (hunk 6 above)
Comment (tezc) [reply src/redisraft.c]: This comment is still valid.

File: src/multi.c (hunk 7 above)
Comment (fadidahanna) [reply src/multi.c:37]: Agree with @sjpotter, I think we should separate logic into different files, as long as we can.

**Review by fadidahanna — COMMENTED** [review]
Still in the middle of reviewing - will continue later on today

File: src/multi.c [hunk 12]
```diff
@@ -77,81 +54,72 @@ bool MultiHandleCommand(RedisRaftCtx *rr,
...
         RaftRedisCommandArrayFree(cmds);
-        RaftRedisCommandArrayMove(cmds, &multiState->cmds);
-        MultiFreeClientState(rr, client_id);
+        RaftRedisCommandArrayMove(cmds, &clientState->multi_state.cmds);
+        ResetMultiClientState(clientState);
 
         return false;
     } else if (cmd_len == 7 && !strncasecmp(cmd_str, "DISCARD", 7)) {
-        if (!multiState) {
+        if (!clientState || !clientState->multi_state.active) {
```
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/multi.c:97]: old code, before we created it on each connection.  to remove.

File: src/redisraft.c (hunk 6 above)
Comment (sjpotter) [reply src/redisraft.c]: right, I didn't remove it, because felt it needed discussion.  outdated is because I had to rebase.

File: src/redisraft.c (hunk 11 above)
Comment (sjpotter) [reply src/redisraft.c:602]: so, I'm not 100% sure about this, while they can be combined, they really are doing 2 different things.  1 is intercepting the asking command (and now that I think of it, should be part of the intercept function), the other is changing the current state of this command due to client state.

Comment (sjpotter) [reply src/redisraft.c:602]: I actually took my idea right here and re factored this code a bit.

File: src/redisraft.c [hunk 13]
```diff
@@ -549,11 +549,60 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
     return false;
 }
 
+static bool AskingHandleCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, RaftRedisCommandArray *cmds)
+{
+    if (cmds->len != 1) {
+        return false;
+    }
+
+    /* TODO: will have to be == 2 in future with magic value in asking for argv[2] */
```
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/redisraft.c:558]: https://github.com/RedisLabs/redisraft/issues/371

File: src/redisraft.c [hunk 14]
```diff
@@ -549,11 +549,60 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
+    size_t cmd_len;
+    const char *cmd_str = RedisModule_StringPtrLen(cmd->argv[0], &cmd_len);
+
+    if (cmd_len != 6 || strncasecmp(cmd_str, "ASKING", 6) != 0) {
+        return false;
+    }
+
+    Node *ret;
+    if (checkLeader(rr, ctx, &ret) == RR_ERROR) {
+        /* change to replyAsk once that's committed */
```
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/redisraft.c]: so, I tried to change it, can't seem to change it "correctly" so just replying moved to it for now.

File: src/redisraft.c [hunk 15]
```diff
@@ -549,11 +549,60 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
+        replyRedirect(ctx, -1, &ret->addr);
+        return true;
+    }
+
+    ClientState *clientState = ClientStateGet(rr, ctx);
+    clientState->asking = true;
```
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/redisraft.c]: unsure if there's much value to that?

File: src/redisraft.c (hunk 11 above)
Comment (sjpotter) [reply src/redisraft.c]: done

File: src/redisraft.c [hunk 16]
```diff
@@ -574,6 +594,14 @@ static void handleRedisCommand(RedisRaftCtx *rr,
         return;
     }
 
+    /* update cmds array with the client's saved "asking" state */
+    cmds->asking = getAskingState(rr, ctx);
```
Comment (tezc) [inline src/redisraft.c:598]: I realized one thing, what is our "asking" policy about the proxy? If these commands come via proxy, "asking" might be true here already and we'll overwrite it back to "false". This might be okay I guess, in the worst case is the request will be rejected. So, either we need a comment about proxy or we should not overwrite this value if is already set? Something like:

```c
/* Comment describing proxy case */
if (!cmds->asking) {
     cmds->asking = getAskingState(rr, ctx);
}
```

File: src/redisraft.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (fadidahanna) [reply src/redisraft.c]: Well, if we have a getter `getAskingState` it makes sense to also have a setter, and keep `asking` field as a private

File: src/redisraft.c [hunk 17]
```diff
@@ -640,6 +652,29 @@ static void handleRedisCommand(RedisRaftCtx *rr,
...
+     * if they have to.
+     */
+    if (handleInterceptedCommands(rr, ctx, cmds)) {
+        return;
+    }
+
+    /* update cmds array with the client's saved "asking" state */
+    if (!cmds->asking) {
+        cmds->asking = getAskingState(rr, ctx);
+    }
```
[the remark this answers is not in the document]
Comment (tezc) [reply src/redisraft.c:668]: No, I mean `cmds->asking = getAskingState(rr,ctx);` line should be enough without the if check.

File: src/multi.c [hunk 18]
```diff
@@ -34,124 +34,82 @@
...
-    bool error;
-} MultiState;
-
-void MultiInitClientState(RedisRaftCtx *rr)
-{
-    rr->multi_client_state = RedisModule_CreateDict(NULL);
-}
-
-uint64_t MultiClientStateCount(RedisRaftCtx *rr)
+void MultiClientStateReset(ClientState * clientState)
```
Comment (tezc) [inline src/multi.c:37]:
```suggestion
void MultiClientStateReset(ClientState *clientState)
```

File: src/redisraft.c [hunk 19]
```diff
@@ -509,6 +509,19 @@ static void handleMigrateCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, RaftRedi
     raft_entry_release(entry);
 }
 
+static void  handleAsking(RedisRaftCtx *rr, RedisModuleCtx *ctx)
```
Comment (tezc) [inline src/redisraft.c:512]:
```suggestion
static void handleAsking(RedisRaftCtx *rr, RedisModuleCtx *ctx)
```

File: src/redisraft.c (hunk 6 above)
Comment (sjpotter) [reply src/redisraft.c]: removed as discussed in zoom meeting

File: src/redisraft.c [hunk 20]
```diff
@@ -509,6 +509,19 @@ static void handleMigrateCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, RaftRedi
...
+{
+    Node *ret;
+    if (checkLeader(rr, ctx, &ret) == RR_ERROR) {
+        replyRedirect(ctx, -1, &ret->addr);
+    }
```
Comment (tezc) [inline src/redisraft.c:517]:
```suggestion
    if (checkLeader(rr, ctx, NULL) == RR_ERROR) {
        return;
    }
```

This should be this way, right? `checkLeader()` already replies with "redirect".

File: src/redisraft.h [hunk 21]
```diff
@@ -797,8 +812,8 @@ long long int RaftLogRewrite(RedisRaftCtx *rr, const char *filename, raft_index_
 void RaftLogRemoveFiles(const char *filename);
 void RaftLogArchiveFiles(RedisRaftCtx *rr);
 RRStatus RaftLogRewriteSwitch(RedisRaftCtx *rr, RaftLog *new_log, unsigned long new_log_entries);
-int RaftMetaRead(RaftMeta *meta, const char* filename);
-int RaftMetaWrite(RaftMeta *meta, const char* filename, raft_term_t term, raft_node_id_t vote);
+int RaftMetaRead(RaftMeta *meta, const char *filename);
+int RaftMetaWrite(RaftMeta *meta, const char*filename, raft_term_t term, raft_node_id_t vote);
```
Comment (tezc) [inline src/redisraft.h:815]: These lines are unrelated but I see you fixed '*' alignment in the first line but you couldn't fix it on the second line :D

File: src/redisraft.c (hunk 20 above)
Comment (sjpotter) [reply src/redisraft.c:517]: actually, not, when passed with &ret, it doesn't redirect (i.e. for proxy mode), but need to remember what I was doing here, as you're right, if redirecting, no need for that.  I think think the point was that this in an ASKING handling, and hence we might want to reply with an ask response and hence would need the leader.  In our case, we don't right now (ask doesn't work for us right now, as the ASK doesn't include the slot we are asking about and hence can't return a well formed ask response, as an aside, part of me thinks that's also something missing from the protocol, an "ASK" response includes the slot its about, an ASKING should include it (and only work on said slot)

File: src/redisraft.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/redisraft.c]: ok, understood now, updated.

File: src/redisraft.h (hunk 21 above)
Comment (sjpotter) [reply src/redisraft.h:815]: I think I just noticed the first one for some reason, so got to it as a nit.

File: src/redisraft.c [hunk 22]
```diff
@@ -509,6 +509,34 @@ static void handleMigrateCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, RaftRedi
...
+static void handleAsking(RedisRaftCtx *rr, RedisModuleCtx *ctx)
+{
+    Node *leader_proxy = NULL;
+
+    Node **ret = rr->config->follower_proxy ? &leader_proxy : NULL;
+    if (checkLeader(rr, ctx, ret) == RR_ERROR) {
+        if (!rr->config->follower_proxy) {
+            return;
+        }
+    }
```
Comment (tezc) [inline src/redisraft.c:533]: This check won't work, please see the usage in `handleRedisCommand()`.

Other than that, probably we don't even need to handle this. Is there a reason we are handling this? We can just set the client "asking" whether we are leader or not.

File: src/util.c [hunk 23]
```diff
@@ -500,4 +500,29 @@ RRStatus parseHashSlots(char * slots, char * string)
 exit:
     RedisModule_Free(string);
     return ret;
-}
\ No newline at end of file
+}
+
+ClientState *ClientStateGet(RedisRaftCtx *rr, RedisModuleCtx * ctx)
```
Comment (tezc) [inline src/util.c:505]:
```suggestion
ClientState *ClientStateGet(RedisRaftCtx *rr, RedisModuleCtx *ctx)
```
There were two misaligned pointer in this line, the previous suggestion had both. Consider accepting changes via github, you don't miss the changes and it's just one click. I find it easier than manual commit.

File: src/redisraft.c (hunk 22 above)
Comment (sjpotter) [reply src/redisraft.c:533]: so, the Q is, if we are not leader, how should we handle an ASKING.

do we 

a) accept it and handle leader later on a "real" command
b) reject it and reply moved back to migrating shardgroup
c) reject it and reply moved to our leader
d) reject it and reply ask to our leader

you're arguing for a. 

c. I'm not sure makes sense because the asking won't transfer with it.  (clients expect that a moved response means they are being redirected to primary owner, i.e. the migrating shard) 
b. is impossible because we dont know who it is (i.e. asking doesn't give us a slot)
d. would be the "best" (i.e. you're asking us, but you want our leader).  But unsure we can do it right (as dont know slot).

now if we do it later, we have 2 options

1) replyMoved back to migrating shard
2) since we know that we are an asking, replyAsk as appropriate (with correct slot info) to our leader.  I'm a bit skeptical on this.

but I'm ok with your suggestion, just wanted to enumerate everything.

Comment (sjpotter) [reply src/redisraft.c:533]: another approach would be to try treat an ASK to a follower as an inherent follower proxy mode?  i.e. we will proxy it to right node, as we can't "nicely" redirect you to it.

Comment (tezc) [reply src/redisraft.c:533]: As I said, if this scenario is super rare, we shouldn't optimize for it. What happens if we don't do this? If this doesn't cause any visible improvement, I'd say we shouldn't write code for it.

Comment (sjpotter) [reply src/redisraft.c:533]: I'm not sure if its super rare, it be behavior I'd expect to see in real life usage.  I actually think as long as we have a proxy mode, it makes sense to leverage it for asking.

Comment (sjpotter) [reply src/redisraft.c:533]: i.e. with the recent understanding that asking resets the state after every execution, every asking will be an independent command returned from an ask response.

with proxy we avoid a lot of ping ponging and just handle it naturally.  the reason to not support proxy normally is to teach the client where to go so it we can avoid the man in the middle.  there wont be ping poinging.  However, for ask/asking thats irrelvant.

With that said, can discuss it tomorrow

File: src/redisraft.c [hunk 24]
```diff
@@ -679,6 +679,11 @@ static void handleRedisCommand(RedisRaftCtx *rr,
         return;
     }
 
+    /* reset asking state when we go to execute the command */
```
Comment (fadidahanna) [inline src/redisraft.c:676]: You should reset the asking state *after* handling the current command.

File: src/redisraft.c [hunk 25]
```diff
@@ -596,14 +586,18 @@ static void handleRedisCommandAppend(RedisRaftCtx *rr,
     }
 
     /* Confirm that we're the leader and handle redirect or proxying if not. */
-    Node **ret = rr->config->follower_proxy ? &leader_proxy : NULL;
+    Node **ret = rr->config->follower_proxy || cmds->asking ? &leader : NULL;
```
Comment (fadidahanna) [inline src/redisraft.c:589]: I suggest to change the `checkLeader` behaviour to check if we are the leader, or if ret exists fill it with value - without reply anything + having a wrapper method that based on the asking status or follower_proxy value reply :
!leader && asking -> reply ASK, 
!leader && !follower_proxy-> reply redirect)
I feel it's hard to follow the logic in the current implementation.

File: src/redisraft.c (hunk 24 above)
Comment (tezc) [reply src/redisraft.c:676]: Also, line 669 and line 677 contradict each other?

File: src/raft.c [hunk 26]
```diff
@@ -187,9 +187,11 @@ static ShardGroup *GetSlotShardGroup(RedisRaftCtx *rr, int slot, bool asking)
 }
 
 static RRStatus validateRaftRedisCommandArray(RedisRaftCtx *rr, RedisModuleCtx *reply_ctx,
-                                              RaftRedisCommandArray *cmds, unsigned int slot)
+                                              RaftRedisCommandArray *cmds)
 {
-    RedisModule_Assert(slot <= REDIS_RAFT_HASH_MAX_SLOT);
+    RedisModule_Assert(0 <= cmds->slot && cmds->slot <= REDIS_RAFT_HASH_MAX_SLOT);
+
+    unsigned int slot = (unsigned int) cmds->slot;
```
[the remark this answers is not in the document]
Comment (tezc) [reply src/raft.c:194]: `-1` is used as logical "no keys".  Maybe, we can define `NO_KEYS = -1` and use the variable unsigned. Then, check like `if (slot == NO_KEYS)`;

File: src/redisraft.c [hunk 27]
```diff
@@ -575,14 +586,18 @@ static void handleRedisCommand(RedisRaftCtx *rr,
...
     if (checkLeader(rr, ctx, ret) == RR_ERROR) {
         return;
     }
 
     /* Proxy */
-    if (leader_proxy) {
-        if (ProxyCommand(rr, ctx, cmds, leader_proxy) != RR_OK) {
+    if (leader) {
+        if (cmds->asking) {
+            if (computeHashSlot(rr, ctx, cmds) == RR_OK){
```
Comment (tezc) [inline src/redisraft.c:597]: What do we do if `computeHashSlot()` returns an error? We have to reply.

Comment (fadidahanna) [reply src/redisraft.c:597]: `computeHashSlot` was changed to support replying in error, but I think we should avoid that.

Comment (tezc) [reply src/redisraft.c:597]: Ok, I see this branch lags behind the master. @sjpotter has already changed this behavior: https://github.com/RedisLabs/redisraft/pull/331

File: src/raft.c (hunk 26 above)
[the remark this answers is not in the document]
Comment (fadidahanna) [reply src/raft.c:194]: actually, we don't use slot value for validation at all, since we return an error from `computeHashSlot` - `-1` is used only to mark that we are handling the first key in `computeHashSlot` - so I guess we can use a local flag for that...

File: src/redisraft.c (hunk 27 above)
Comment (fadidahanna) [reply src/redisraft.c:597]: Well, I'm not sure, because it's part of the changes in the PR... anyway, we all agree that `computeHashSlot` should not reply in case of error... 🙏

File: src/redisraft.c [hunk 28]
```diff
@@ -673,12 +675,54 @@ static void handleRedisCommand(RedisRaftCtx *rr,
...
+     * commands have no keys, which is something Redis Cluster never does. We
+     * still need to consider how this impacts clients which may not expect it.
+     */
+    replyRedirect(ctx, 0, &leader->addr);
+}
+
+static RRStatus handleNonLeaderCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader, RaftRedisCommandArray *cmds) {
+    RedisModule_Assert(!raft_is_leader(rr->raft));
+
+    // reply ASK for ASKING state commands
```
Comment (tezc) [inline src/redisraft.c:700]: I guess this is one of our rare "consistent" coding convention, we use `/* */` for comments.

File: src/redisraft.c [hunk 29]
```diff
@@ -673,12 +675,54 @@ static void handleRedisCommand(RedisRaftCtx *rr,
...
+    if (cmds->asking) {
+        RedisModule_Assert(cmds);
```
Comment (tezc) [inline src/redisraft.c:702]: Assert won't be triggered ever as we deference it on line 701.

File: src/redisraft.c [hunk 30]
```diff
@@ -673,12 +675,54 @@ static void handleRedisCommand(RedisRaftCtx *rr,
...
+            return RR_ERROR;
+        }
+
+        replyAsk(rr, ctx, (unsigned int) slot);
+        return RR_OK;
+    }
+
+    // forward commands to the leader in follower_proxy state
+    if (rr->config->follower_proxy) {
+        RedisModule_Assert(cmds);
```
Comment (tezc) [inline src/redisraft.c:715]: Same here, line 701 deference it already.

File: src/commands.c [hunk 31]
```diff
@@ -241,3 +242,19 @@ unsigned int CommandSpecGetAggregateFlags(RaftRedisCommandArray *array, unsigned
...
+
+bool IsKeyCommands(RedisRaftCtx *rr,
+                   RedisModuleCtx *ctx,
+                   RaftRedisCommandArray *cmds) {
+    for (int i = 0; i < cmds->len; i++) {
+        RaftRedisCommand *cmd = cmds->commands[i];
+
+        /* Iterate command keys */
+        int num_keys = 0;
+        RedisModule_GetCommandKeys(rr->ctx, cmd->argv, cmd->argc, &num_keys);
```
Comment (tezc) [inline src/commands.c:254]: This functions returns something and you need to free it if I'm not mistaken.

File: src/raft.c [hunk 32]
```diff
@@ -272,20 +270,23 @@ static RRStatus validateRaftRedisCommandArray(RedisRaftCtx *rr, RedisModuleCtx *
...
+    unsigned int slot;
+
     if (!isSharding(rr)) {
         return RR_OK;
     }
 
-    if (computeHashSlot(rr, ctx, cmds) != RR_OK) {
-        return RR_ERROR;
+    /* If commands have no keys, continue */
+    if (!IsKeyCommands(rr, ctx, cmds)) {
```
Comment (tezc) [inline src/raft.c:280]: I think we duplicate the work here. We loop over the command array and pull the keys. We are going to do the same thing in `computeHashSlot()`. I know we did this to avoid cast for `slot` but I'm not sure it is worth the effort. Maybe, we should just go with cast?

File: src/common.c [hunk 33]
```diff
@@ -127,47 +127,32 @@ raft_node_t *getLeaderNodeOrReply(RedisRaftCtx *rr, RedisModuleCtx *ctx)
...
 
-    /* One anomaly here is that may redirect a client to the leader even for
-     * commands have no keys, which is something Redis Cluster never does. We
-     * still need to consider how this impacts clients which may not expect it.
-     */
-    replyRedirect(ctx, 0, &node->addr);
-    return RR_ERROR;
+
+/* Checks if the current cluster is the leader */
+bool isLeader(RedisRaftCtx *rr) {
```
Comment (tezc) [inline src/common.c:153]: Sorry for nitpick, as this function doesn't do anything on top of `raft_is_leader()`, probably we don't need it.

File: src/redisraft.c (hunk 29 above)
Comment (fadidahanna) [reply src/redisraft.c:702]: yes, this is a leftover from previous impl. fixed.

File: src/redisraft.c [hunk 34]
```diff
@@ -645,6 +661,72 @@ static void handleRedisCommand(RedisRaftCtx *rr,
     raft_entry_release(entry);
 }
 
+static void handleRedisCommand(RedisRaftCtx *rr,
+                               RedisModuleCtx *ctx, RaftRedisCommandArray *cmds) {
```
Comment (tezc) [inline src/redisraft.c:665]:
```suggestion
static void handleRedisCommand(RedisRaftCtx *rr,
                               RedisModuleCtx *ctx, RaftRedisCommandArray *cmds) 
{
```

File: src/redisraft.c [hunk 35]
```diff
@@ -645,6 +661,72 @@ static void handleRedisCommand(RedisRaftCtx *rr,
...
+    RedisModule_Assert(!raft_is_leader(rr->raft));
+    RedisModule_Assert(cmds);
+
+    /* reply ASK for ASKING state commands */
+    if (cmds->asking) {
+        int slot;
+        if (computeHashSlot(rr, ctx, cmds, &slot) != RR_OK) {
+            return RR_ERROR;
+        }
+        RedisModule_Assert(slot != -1); /* ASKING mode requests always have keys */
```
Comment (tezc) [inline src/redisraft.c:711]: I think we should not crash even if a client sends an "asking" command without a key. Client might be buggy or it might be in the development phase etc. We should reply an error in this case.

File: src/redisraft.c [hunk 36]
```diff
@@ -645,6 +661,72 @@ static void handleRedisCommand(RedisRaftCtx *rr,
...
+
+static RRStatus handleNonLeaderCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader, RaftRedisCommandArray *cmds) {
+    RedisModule_Assert(!raft_is_leader(rr->raft));
+    RedisModule_Assert(cmds);
+
+    /* reply ASK for ASKING state commands */
+    if (cmds->asking) {
+        int slot;
+        if (computeHashSlot(rr, ctx, cmds, &slot) != RR_OK) {
+            return RR_ERROR;
```
Comment (tezc) [inline src/redisraft.c:709]: We should also reply an error here.

Comment (tezc) [reply src/redisraft.c:709]: I guess we can do something like: 
```
if slot = -1: "ERR asking without keys"
if not RR_OK: replyCrossSlot();
```

File: src/redisraft.c [hunk 37]
```diff
@@ -645,6 +661,72 @@ static void handleRedisCommand(RedisRaftCtx *rr,
...
+    if (MultiHandleCommand(rr, ctx, cmds)) {
+        return;
+    }
+
+    handleRedisCommandAppend(rr, ctx, cmds);
+
+    /* reset asking state when after executing the command */
+    if (cmds->asking) {
+        setAskingState(rr, ctx, false);
+    }
```
Comment (tezc) [inline src/redisraft.c:687]:
```suggestion
    setAskingState(rr, ctx, false);
```
I think we don't need an if check here.

File: src/redisraft.c [hunk 38]
```diff
@@ -645,6 +661,72 @@ static void handleRedisCommand(RedisRaftCtx *rr,
...
+}
```
Comment (tezc) [inline src/redisraft.c:688]:
```suggestion
    /* update cmds array with the client's saved "asking" state */
    cmds->asking = getAskingState(rr, ctx);
    setAskingState(rr, ctx, false);

    if (handleInterceptedCommands(rr, ctx, cmds)) {
        return;
    }

    /* Check if MULTI/EXEC bundling is required. */
    if (MultiHandleCommand(rr, ctx, cmds)) {
        return;
    }

    handleRedisCommandAppend(rr, ctx, cmds);
}
```

I guess we should clear "asking" state no matter what.

File: src/redisraft.c [hunk 39]
```diff
@@ -639,10 +656,83 @@ static void handleRedisCommand(RedisRaftCtx *rr,
...
+        int slot;
+        RRStatus res = computeHashSlot(rr, ctx, cmds, &slot);
+        if (slot == -1) {
+            /* ASKING mode requests should always have keys */
+            RedisModule_ReplyWithError(ctx, "ERR cmd without keys in asking mode");
+            return RR_ERROR;
+        }
+
+        if (res == RR_ERROR) {
+            replyCrossSlot(ctx);
```
Comment (tezc) [inline src/redisraft.c:717]:
```suggestion
            replyCrossSlot(ctx);
            return RR_ERROR;
```
We need to return here. 
Btw looks like this function does not need a return value.

Comment (fadidahanna) [reply src/redisraft.c:717]: I prefer to have it right now, it may be useful for the error handling task.

Comment (tezc) [reply src/redisraft.c:717]: sure

**Review by tezc — APPROVED** [review]
